### PR TITLE
feat(treesitter)!: make correct match behavior for quantified captures the default

### PR DESCRIPTION
**Pull request by gpanders**
For context, see https://github.com/neovim/neovim/pull/24738. Before that PR, Nvim did not correctly handle captures with quantifiers. That PR made the correct behavior opt-in to minimize breaking changes, with the intention that the correct behavior would eventually become the default. Users can still opt-in to the old (incorrect) behavior for now, but this option will eventually be removed completely.

BREAKING CHANGE: Any plugin which uses `Query:iter_matches()` must update their call sites to expect an array of nodes in the `match` table, rather than a single node.

## Migration guide

**Before**

```lua
for pattern, match, metadata in query:iter_matches(...) do
  for id, node in pairs(match) do
    -- Do something with node
  end
end
```

**After**

```lua
for pattern, match, metadata in query:iter_matches(...) do
  for id, nodes in pairs(match) do
    for _, node in ipairs(nodes) do
      -- Do something with node
    end
  end
end
```

### CODE REVIEW

**Review by clason — APPROVED** [review]
Change LGTM (should not be breaking to downstream as of now), but documentation is a bit wordy maybe and could be boiled down? (I don't think we need to call out the specific version in the docstring, for example; that's what `news.txt` is for.)

**Comment by gpanders** [comment]
>I don't think we need to call out the specific version in the docstring, for example; that's what news.txt is for.

The old docstring did that but was removed in this PR. I admit the news blurb is a bit wordy, but I'm not sure how to pare it down without either reducing clarity or eliminating remediation steps (which is valuable for breaking change notes).

File: runtime/doc/news.txt [hunk 1]
```diff
@@ -89,7 +89,13 @@ PLUGINS
 
 TREESITTER
 
-• TODO
+• |Query:iter_matches()| now defaults to returning all matching nodes in a
+  match, rather than simply the last node in a match. In particular, the
```
Comment (MariaSolOs) [inline runtime/doc/news.txt:93]: Nit, but I think this bit is redundant:
```suggestion
  match, rather than simply the last node. In particular, the
```

File: runtime/doc/treesitter.txt [hunk 2]
```diff
@@ -1252,8 +1244,9 @@ Query:iter_matches({node}, {source}, {start}, {stop}, {opts})
                   • all (boolean) When set, the returned match table maps
                     capture IDs to a list of nodes. Older versions of
                     iter_matches incorrectly mapped capture IDs to a single
-                    node, which is incorrect behavior. This option will
-                    eventually become the default and removed.
+                    node, which is incorrect behavior. As of Nvim 0.11 this
```
Comment (clason) [inline runtime/doc/treesitter.txt:1247]: I was referring to this bit here ☝️

Comment (gpanders) [reply runtime/doc/treesitter.txt]: Thanks, updated.

File: runtime/doc/news.txt (hunk 1 above)
Comment (gpanders) [reply runtime/doc/news.txt]: Changed as suggested.

File: runtime/lua/vim/treesitter/query.lua [hunk 3]
```diff
@@ -925,9 +919,9 @@ end
 ---   - max_start_depth (integer) if non-zero, sets the maximum start depth
 ---     for each match. This is used to prevent traversing too deep into a tree.
 ---   - match_limit (integer) Set the maximum number of in-progress matches (Default: 256).
----   - all (boolean) When set, the returned match table maps capture IDs to a list of nodes.
----     Older versions of iter_matches incorrectly mapped capture IDs to a single node, which is
----     incorrect behavior. This option will eventually become the default and removed.
+---   - all (boolean) When set (the default), the returned match table maps capture IDs to a list of
+---     nodes. Older versions of iter_matches incorrectly mapped capture IDs to a single node, which
+---     is incorrect behavior. This option will be removed in a future release.
```
Comment (clason) [inline runtime/lua/vim/treesitter/query.lua:924]:
```suggestion
--- - all (boolean) When `false` (default `true`), the returned table maps capture IDs to a single
---   (last) node instead of the full list of matching nodes. This option is only for backward 
---   compatibility and will be removed in a future release.
```

File: runtime/doc/news.txt [hunk 4]
```diff
@@ -89,7 +89,13 @@ PLUGINS
...
 TREESITTER
 
-• TODO
+• |Query:iter_matches()| now defaults to returning all matching nodes in a
+  match, rather than simply the last node. In particular, the `match` table
+  returned by |Query:iter_matches()| now maps capture IDs to a list of nodes,
+  rather than a single node. The old behavior is incorrect, but can be opted
+  into by setting `all=false` in the `opts` argument to
+  |Query:iter_matches()|. This option will be removed completely in a future
+  release.
```
Comment (clason) [inline runtime/doc/news.txt:98]:
```suggestion
• |Query:iter_matches()| correctly returns all matching nodes in a match
  instead of only the last node. This means that the returned table maps
  capture IDs to a list of nodes that need to be iterated over. For
  backwards compatibility, an option `all=false` (only return the last
  matching node) is provided that will be removed in a future release.
```

**Review by clason — COMMENTED** [review]
Suggestions for condensed text; up to you if you want to include (parts of) them; feel free to merge then.

File: runtime/lua/vim/treesitter/query.lua (hunk 3 above)
Comment (gpanders) [reply runtime/lua/vim/treesitter/query.lua]: Changed as suggested.

File: runtime/doc/news.txt (hunk 4 above)
Comment (gpanders) [reply runtime/doc/news.txt]: Changed as suggested.

**Comment by glepnir** [comment]
after this patch `TSNode:parent` can be nil ..

```
Error in decoration provider treesitter/highlighter.line:
Error executing lua: ...nvim-treesitter/lua/nvim-treesitter/query_predicates.lua:80: attempt to call method 'parent' (a nil value)
````

**Comment by clason** [comment]
Nvim-treesitter problem; looks like the (compatibility) predicate needs to be adapted (or, better yet, removed).

**Comment by glepnir** [comment]
aha sry for noise then hide comment ..
